lib/block-supports/layout.php

```diff
@@ -93,6 +93,7 @@ function gutenberg_get_layout_style( $selector, $layout, $has_block_gap_support
 		if ( $has_block_gap_support ) {
 			$gap_style = $gap_value ? $gap_value : 'var( --wp--style--block-gap, 0.5em )';
 			$style    .= "gap: $gap_style;";
+			$style    .= "--wp--style--scoped-block-gap: $gap_style;";
```

Comment:
Interesting approach, thanks, @glendaviesnz !

I see the need for block scoping in relation to gap support, and I believe it might help avoid the issues we saw in https://github.com/WordPress/gutenberg/issues/36521

The only thing I'd raise as a potential issue is the coupling with the layout abstraction, that is, the requirement to opt-in to `__experimentalLayout`. 

If we look at this is as a feature to be used elsewhere, and not just the Gallery, I'd imagine some blocks might want to take advantage of a scope block gap CSS var, but I can also imagine that not all blocks would want to have the extra styles associated with the layout rules.

Not only that, I'd be interested to know whether it's sustainable to require theme developers to know that two support rules are required for scoped blockGap to work.

I don't know if these are legitimate concerns. The `gap` rule is a property of grid, flex, and multi-column layouts, which are..., well, "layouts" so there is that 😄

Reply:
Good questions - my reason for putting it there was that currently, you can't get `gap` support at all unless you opt into `layout`. It actually took a bit of time to get this PR started as I couldn't work out why the gap support wasn't showing when I had opted into `spacing` 🤔. 

I am guessing to move `--wp--style--scoped-block-gap` out of `layout`  would require also moving `gap`, which I assume would take a lot of rehashing of some previous decisions about why this is best placed in `layout`

Layout does have [some limitations I had to override in terms of the gallery](https://github.com/WordPress/gutenberg/pull/37903/files#diff-41629d44df4cc8609f305c127f4f22111c6853e357645e1b3eb88a1d906360b0R6), but I think these can be sorted in separate PRs.

Reply:
For the record - I'm having a related intellectual struggle about the role of layout in another PR: **Experimenting with adding writing mode to layout** #37932 

There's an interesting discussion about a "[styles engine](https://github.com/WordPress/gutenberg/discussions/37495)", which, one day, might help with issues such as these. 

I don't want to raise it as a panacea – we have immediate needs that might, for now, require some compromises such as the ones we're introducing in this PR – but I can easily envision that a centralized rendering engine might be the brain butter we need when making such decisions!

Reply:
Hi There! There are indeed interesting challenges here and I recommend you read through this PR a bit to understand the decisions about the coupling #37360 and why we want to avoid CSS variables as well.

--
Basically I have several concerns about this code:

 - Conceptually, I think ideally the layout should provide everything and blocks shouldn't need to access the block gap. It seems that's not possible here according to your PR description (I'd like to try to understand this more and see if we can find an alternative).
- The question that could be asked with the current implementation is why we need the variable for "flex" layout but not other kind of layouts (we have "flow" as well), and the answer is not clear aside the fact that this implemented to solve a specific case (Gallery block uses flex) and adding the CSS variable in all layouts is a no-go (see linked issue/PR)

--
If we're not able to find any solution without involving the CSS variable, I'd recommend a specific hook for the Gallery block to access its attribute (block gap) and generate the required styles without variables and independently of the "layout" block support.

Reply:
Thanks, @youknowriad, that is useful feedback. 

I think I have explored all the options for solving this without a css var, and haven't found a suitable alternative due to the fact that we have a variable number of columns and images, and require the uneven images to equally distribute on the last row, and due to the fact that the key style that needs to update is the width of the innerBlocks nested images, not the parent gallery block styles ... but I will take a look at using a custom hook and looking for a gallery specific solution rather than trying to solve this within the generic layout code.